src/ui_model/model_layout.rs

```diff
@@ -6,11 +6,28 @@ use unicode_width::UnicodeWidthStr;
 use crate::highlight::Highlight;
 use crate::ui_model::UiModel;
 
+#[derive(Clone)]
+pub struct HighlightedRange {
+    pub highlight: Rc<Highlight>,
+    pub graphemes: Vec<String>,
```

Comment:
this is a bit a weird data structure to have one string for each grapheme. Probably, one could use the grapheme offsets + original string or use crates like smolstr, that would enable small string optimizations to avoid allocating for each character on screen. but since current code seems to work :man_shrugging:

Reply:
afaict when I've profiled neovim-gtk (at least on fedora 36) it's usually pango rendering that takes the longest, followed by pretty much all of this ui model code. Granted, this code is usually far less of a bottleneck (and the biggest performance improvement we could probably get is from improving our pango itemizer) - but optimizing stuff here probably isn't a bad idea.